wallet-new/test/unit/UTxO/BlockGen.hs

```diff
@@ -219,3 +236,175 @@ zipFees
+  , _treeGenGlobalCtxMaxHeight = 50
+  , _treeGenGlobalCtxSharedTransactionLikelihood = 0.5
+  , _treeGenGlobalCtxBootTransaction = boot
+  }
+
+data TreeGenBranchCtx h = TreeGenBranchCtx
+    { _treeGenBranchCtxPrincipalBranch :: !Bool
+      -- ^ Is this the principal branch? The principal branch will never
+      --   be pruned.
+    , _treeGenBranchCtxBranchHeight    :: !Int
+      -- ^ Height of the tip of the branch (from the root).
+    , _treeGenBranchCtxCurrentUtxo     :: !(Utxo h Addr)
+      -- ^ The mapping of current addresses and their current account values.
+    , _treeGenBranchCtxTransactions    :: ![Value -> Transaction h Addr]
+      -- ^ Transactions to form this block. This is stored here because we
+      --   generate all transactions for the child block at the parent level,
+      --   in order to facilitate sharing.
+    }
+
+makeFields ''TreeGenBranchCtx
+
+-- | Tree generation monad.
+newtype TreeGen h a = TreeGen
+    { unTreeGen :: StateT (TreeGenGlobalCtx h) Gen a
+    } deriving (Functor, Applicative, Monad, MonadState (TreeGenGlobalCtx h))
+
+-- | Lift a 'Gen' action into the 'TreeGen' monad.
+liftGenTree :: Gen a -> TreeGen h a
+liftGenTree = TreeGen . lift
+
+genValidBlocktree :: Hash h Addr => PreTree h Gen ()
+genValidBlocktree = toPreTreeWith identity newTree
+
+toPreTreeWith
+    :: Hash h Addr
+    => (TreeGenGlobalCtx h -> TreeGenGlobalCtx h) -- ^ Modify the global settings
+    -> TreeGen h (NoFeeBlockTree h)
+    -> PreTree h Gen ()
+toPreTreeWith settings bg = DepIndep $ \(boot :: Transaction h Addr)-> do
+    ks <- evalStateT (unTreeGen bg) (settings (initTreeGenGlobalCtx boot))
+    return $ \fees ->
+      (OldestFirst (fmap (OldestFirst . dropWhile (== boot))
+                         (zipTreeFees ks fees)), ())
+
+newTree :: forall h. Hash h Addr
+        => TreeGen h (NoFeeBlockTree h)
+newTree = do
+    boot <- use bootTransaction
+    -- Choose a random height for the blocktree
+    height <- liftGenTree $ choose (10, 50)
+    maxHeight .= height
+    Tree.unfoldTreeM buildTree $ initBranchCtx boot
+  where
+    initBranchCtx boot = TreeGenBranchCtx True 0 (trUtxo boot) [const boot]
+    buildTree :: TreeGenBranchCtx h
+              -> TreeGen h ([Value -> Transaction h Addr], [TreeGenBranchCtx h])
+    buildTree branchCtx = do
+      let curHeight = branchCtx ^. branchHeight
+
+      -- Firstly, decide whether we should prune this branch. We prune if
+      -- - we have reached the maximum height, or
+      -- - we are not the principal branch, and
+      -- - - we have reached the maximum height - 1, or
+      -- - - with probability equal to the prune likelihood
+      toPrune <- do
+        pl <- use pruneLikelihood
+        maxH <- use maxHeight
+        toss <- liftGenTree $ choose (0,1)
+        return $ (curHeight >= maxH)
+               || ((not $ branchCtx ^. principalBranch)
+                   && ((curHeight >= maxH - 1) || toss < pl)
+                  )
```

Comment:
Tricky boolean logic always messes me up. Would it be useful to factor this out into a separate function and write tests on it?